### PR TITLE
DOC: dynamically create a version pull-down using javascript

### PR DESCRIPTION
**Pull request by mattip**
Enhancement to create a version pulldown rather than a link to a static list of `versions.html`

- The version is now passed to sphinx from the environment, so the post-processing to mangle the version is no longer needed
- The layout template has been changed
  - The javascript scripts are loaded in the header, this should be backported to the upstream [pytorch_sphinx_theme repo](https://github.com/pytorch/pytorch_sphinx_theme) since now the loading stanza appears twice on the page :(. I needed this to activate the new javascript function as the page loads.
  - A new snippet of javascript builds a `ul` list of versions on document loading (provides consistency across documentation versions)
    - The snippet will try to match the current document to the one from the other versions. If there is a match, switching versions will keep you at the same page. 
  - The version list is now below the search bar.

I could not get the page existence check to work asynchronously, so page loading time takes a hit. Is the UX improvement worth it? I was basing this off the [warning in the NumPy docs](https://numpy.org/doc/1.13/) when you view older documentation, [this](https://numpy.org/doc/1.13/reference/routines.random.html) is a page with no parallel in the latest version, so the message will pre-populate a search qeury (source code [here](https://github.com/numpy/doc/blob/master/_static/versionwarning.js_t)). maybe I missed something? In any case, the NumPy script only needs to check for the latest page, my adaptaion scans all the versions.

Help here would be welcome from people fluent in web front-end programming.

### CODE REVIEW

File: docs/source/_static/docversions.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+                label += ' (stable release)';
+                break;
+        }
+        if (typeof(DOCUMENTATION_OPTIONS) !== 'undefined') {
+            if (DOCUMENTATION_OPTIONS['VERSION'] == versions[i])
+            {
+                open_list = '<li id="current">'
+            }
+        }
+        const pathPattern = /docs\/(build\/html|stable|master|[0-9.rc]+)(.*)/;
```
Comment (mattip) [inline docs/source/_static/docversions.js:27]: whether you build locally or deploy to pytorch/pytorch.github.io, the document URL will look like
```
file:///home/user/pytorch/docs/build/html/generated/torch.nn.parameter.Parameter.html
or
https://pytorch.org/docs/1.6.0/generated/torch.nn.parameter.Parameter.html
```
So the `docs` part is a consistent anchor, then either `build/html` or one of the versions (`stable', `master`, `1.6.0` etc), then the actual document name we are interested in.

File: docs/source/_static/docversions.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+            url = 'https://pytorch.org/docs/' + versions[i] + m[2];
+            // Checks synchronously if the page exists, fail -> replace the link
+            $.ajax({url: url, cache: true, async: false}).fail(function(jqXHR, textStatus) {
+                url = base_url
+                });
+        }
+        document.write(open_list);
+        document.write('<a id=ID href="URL">VERSION</a> </li>\n'
+                        .replace('VERSION', label)
+                        .replace('URL', url));
```
Comment (mattip) [inline docs/source/_static/docversions.js:42]: I couldn't get this to work with `async: true` and then rewriting the `item.href`.

File: docs/source/_templates-stable/layout.html [hunk 3]
```diff
@@ -1,11 +1,44 @@
...
+{% block extrahead %}
+       <!-- Early load javascript files -->
+       <script type="text/javascript" id="documentation_options" data-url_root="{{ pathto('', 1) }}" src="{{ pathto('_static/documentation_options.js', 1) }}"></script>
+       {%- for scriptfile in script_files %}
+         {{ js_tag(scriptfile) }}
+       {%- endfor %}
+       <script type="text/javascript" data-url_root="{{ pathto('', 1) }}" src="{{ pathto('_static/docversions.js', 1) }}"></script>
+{{ super() }}
+{% endblock %}
+
```
Comment (mattip) [inline docs/source/_templates-stable/layout.html:13]: This si the bit that should be upstreamed to the theme (without the `_static/docversions.js`). Is there a reason they have it [so late in the template](https://github.com/pytorch/pytorch_sphinx_theme/blob/master/pytorch_sphinx_theme/layout.html#L290) there? Connected to `{% if not embedded %}` ?

File: docs/source/_templates-stable/layout.html [hunk 4]
```diff
@@ -1,11 +1,44 @@
...
+                </ul>
+                <script type="text/javascript">
+                  function versiontoggle() {
+                    $("#versionList").toggle()
+                  };
+                </script>
+              {% endif %}
+            {% endif %}
+{% endblock %}
+
```
Comment (mattip) [inline docs/source/_templates-stable/layout.html:41]: This block overrides the [original sidebartitle block](https://github.com/pytorch/pytorch_sphinx_theme/blob/master/pytorch_sphinx_theme/layout.html#L192)

**Comment by mattip** [comment]
Now the linking is working properly. [This page](https://8404202-65600975-gh.circle-artifacts.com/0/docs/data.html) is available in all versions, [this page](https://8404202-65600975-gh.circle-artifacts.com/0/docs/generated/torch.nn.parameter.Parameter.html#torch.nn.parameter.Parameter) is only available in versions 1.6.0 and higher.

**Review by brianjo — APPROVED** [review]
I looked at the preview for this and I think it's good. Would love to see this added to release going forward, and master. (We don't need to rebuild previous versions.)

File: docs/source/_static/docversions.js [hunk 5]
```diff
@@ -0,0 +1,61 @@
+/* Create version-related dynamic output: instert_version_links() will output
```
Comment (seemethere) [inline docs/source/_static/docversions.js:1]:
```suggestion
/* Create version-related dynamic output: insert_version_links() will output
```

Just a nit, but there's a typo here

**Review by seemethere — APPROVED** [review]
looks mostly good, just one minor nit

**Comment by brianjo** [comment]
Converted to Draft, @ranman is going to work on some changes.

**Comment by mattip** [comment]
This was mostly a proof-of-concept. Closing. At some point, the pytorch doc team may want to look at how the [pydata theme](https://pydata-sphinx-theme.readthedocs.io/en/latest/) implemented a version pulldown.

**Pull request closed without merging**